agent/engine/dockerstate/docker_task_engine_state.go

```diff
@@ -457,3 +470,70 @@ func (state *DockerTaskEngineState) GetTaskByIPAddress(addr string) (string, boo
 	taskARN, ok := state.ipToTask[addr]
 	return taskARN, ok
 }
+
+// GenerateAndAddV3EndpointID generates a container identifier to be used in v3 metadata endpoint,
+// and update relevant maps. Return false if we fail to generate an unique identifier, otherwise
+// return true
+func (state *DockerTaskEngineState) GenerateAndAddV3EndpointID(task *apitask.Task, container *apicontainer.Container, uuid uuidwrapper.UUID) bool {
+	state.lock.Lock()
+	defer state.lock.Unlock()
+
+	if container.V3EndpointID != "" { // container's v3 endpoint id has already been set
+		seelog.Debugf("Container's v3 endpoint id has already been set to: %s", container.V3EndpointID)
+		state.v3EndpointIDToTask[container.V3EndpointID] = task.Arn
+		state.v3EndpointIDToContainerName[container.V3EndpointID] = container.Name
+		return true
+	}
+
+	for i := 0; i < maximumGenerateIDRetries; i++ {
+		randID := uuid.New()
+		if _, ok := state.v3EndpointIDToContainerName[randID]; !ok { // we've generated an unique ID
+			seelog.Debugf("Set container's v3 endpoint id to: %s", randID)
+			container.V3EndpointID = randID
+			state.v3EndpointIDToTask[randID] = task.Arn
+			state.v3EndpointIDToContainerName[randID] = container.Name
+			return true
+		}
+	}
+
+	seelog.Debugf("Unable to generate unique v3 endpoint id after %d retries", maximumGenerateIDRetries)
+	return false
+}
+
+// ContainerByV3EndpointID returns an apicontainer.DockerContainer for a given v3 endpoint ID
+func (state *DockerTaskEngineState) ContainerByV3EndpointID(v3EndpointID string) (*apicontainer.DockerContainer, bool) {
+	state.lock.RLock()
+	defer state.lock.RUnlock()
+
+	taskArn, ok := state.v3EndpointIDToTask[v3EndpointID]
+	if !ok {
+		return nil, false
+	}
+
+	containerName, ok := state.v3EndpointIDToContainerName[v3EndpointID]
```

Comment:
what does it mean if we fail here instead of later? are the failure conditions different at each step here?

Reply:
i think currently, if the first check succeeds, we won't fail any of the later checks. it's just a sanity check in case the updating logic related to those maps change in the future.